Core/AI: enable SmartScripts on Scenes + add scene related actions/events

Traesh opened the pull request:
**Changes proposed:**

- Allow SmartScript targeting scenes, with entryOrGuid = sceneId and source_type = 10
- Add new "event_param_string" to smart_script database (usefull by example in SMART_EVENT_SCENE_TRIGGER)

- Add four new SMART_EVENTs :
    - SMART_EVENT_SCENE_START
    - SMART_EVENT_SCENE_TRIGGER
    - SMART_EVENT_SCENE_CANCEL
    - SMART_EVENT_SCENE_COMPLETE


- Add two new SMART_ACTIONs :
    - SMART_ACTION_SCENE_PLAY
    - SMART_ACTION_SCENE_CANCEL

**Target branch(es):** 3.3.5/master

- [ ] 3.3.5
- [x] master

**Issues addressed:** Closes #19189

**Tests performed:** (Does it build, tested in-game, etc.)
Built, tested ingame with following SQL, should spawn creature 1 on scene 1181 triggering event (.cast spell 201179)
```sql
INSERT INTO `smart_scripts` VALUES (1181,10,0,0,79,0,100,0,0,0,0,0,'BURN PLAYER',12,1,1,0,0,0,0,1,0,0,0,0,0,0,0,'');

```

**Known issues and TODO list:** (add/remove lines as needed)

- [ ]

Traesh committed:
Add SMART_ACTION_SCENE_PLAY & SMART_ACTION_SCENE_CANCEL

Aokromes commented on src/server/game/AI/SmartScripts/SmartScriptMgr.h:
```diff
@@ -558,8 +558,19 @@ enum SMART_ACTION
     SMART_ACTION_DISABLE_EVADE                      = 117,    // 0/1 (1 = disabled, 0 = enabled)
-
-    SMART_ACTION_END                                = 118
+    SMART_ACTION_SET_CAN_FLY                        = 119,    // 0/1 (3.3.5 only, NYI)
```
118 to 127 are on used on 3.3.5a.

Traesh replied:
Yup, already saw it in another PR, that's why i set placeholder with comments (3.3.5 only, NYI) :)

Maybe a bad idea, will remove it if required

Traesh committed:
Enable SmartScript on players + add 4 new scene related events

SMART_EVENT_SCENE_START
SMART_EVENT_SCENE_TRIGGER
SMART_EVENT_SCENE_CANCEL
SMART_EVENT_SCENE_COMPLETE

Traesh committed:
Add missing SQL

Shauren commented:
can i just say plsno to player sai?

Traesh commented:
Yup, can i ask why ? :p

i feel like this could be usefull in some case

LuigiElleBalotta commented:
Yes that could be useful

streetrat commented:
this is a little overkill
scene scripting will be needed, but using sai vs players is just bad
problems:
- you cant use any timed stuff, only if you hook into player::update, just like for creatures
- worst: ALL player scripts will be loaded to ALL players... imagine a year later, when all scenes are scripted, and you have 2k sai scripts for those scenes.. you have 1k player with 2k sai.. and probably 99.9999% nothing will be exacuted

if you want to use sai + players + events, then instead of hacking into players, create a new source type (scenes) and use the same hooks as you did, the invoker will be the player, this is how sai should be used

Traesh commented:
@streetrat You got me at "imagine a year later, when all scenes are scripted", you now have one year to script them all :p

I agree on overkill for scenes, i'll switch behavior to a more conventional way

Traesh committed:
Delete Player SAI, switch to Scenes SAI

Traesh committed:
Add default to `event_param_string`

tkrokli commented:
Note to the dev merging this issue in the future: `Closes #19189` is missing in the commit message. This can be added in the github merge dialog (the final commit message needs editing anyway).

Traesh commented:
@Shauren It should be better now, player sai removed :)

Traesh committed:
Merge branch 'master' into scene_smart
